apiserver/facades/agent/uniter/goal-state_test.go

```diff
@@ -0,0 +1,336 @@
+	})
+	s.machine1 = factory.MakeMachine(c, &jujufactory.MachineParams{
+		Series: "quantal",
+		Jobs:   []state.MachineJob{state.JobHostUnits},
+	})
+	s.machine2 = factory.MakeMachine(c, &jujufactory.MachineParams{
+		Series: "quantal",
+		Jobs:   []state.MachineJob{state.JobHostUnits},
+	})
+
+	s.wpCharm = factory.MakeCharm(c, &jujufactory.CharmParams{
+		Name: "wordpress",
+		URL:  "cs:quantal/wordpress-0",
+	})
+	s.wordpress = factory.MakeApplication(c, &jujufactory.ApplicationParams{
+		Name:  "wordpress",
+		Charm: s.wpCharm,
+	})
+	s.wordpressUnit = factory.MakeUnit(c, &jujufactory.UnitParams{
+		Application: s.wordpress,
+		Machine:     s.machine0,
+	})
+
+	mysqlCharm := factory.MakeCharm(c, &jujufactory.CharmParams{
+		Name: "mysql",
+	})
+	s.mysql = factory.MakeApplication(c, &jujufactory.ApplicationParams{
+		Name:  "mysql",
+		Charm: mysqlCharm,
+	})
+	s.mysqlUnit = factory.MakeUnit(c, &jujufactory.UnitParams{
+		Application: s.mysql,
+		Machine:     s.machine1,
+	})
+
+	s.meteredCharm = s.Factory.MakeCharm(c, &jujufactory.CharmParams{
+		Name: "metered",
+		URL:  "cs:quantal/metered",
+	})
+	s.meteredService = s.Factory.MakeApplication(c, &jujufactory.ApplicationParams{
+		Charm: s.meteredCharm,
+	})
+	s.meteredUnit = s.Factory.MakeUnit(c, &jujufactory.UnitParams{
+		Application: s.meteredService,
+		SetCharmURL: true,
+	})
+
+	// Create a FakeAuthorizer so we can check permissions,
+	// set up assuming unit 0 has logged in.
+	s.authorizer = apiservertesting.FakeAuthorizer{
+		Tag: s.mysqlUnit.Tag(),
+	}
+
+	// Create the resource registry separately to track invocations to
+	// Register.
+	s.resources = common.NewResources()
+	s.AddCleanup(func(_ *gc.C) { s.resources.StopAll() })
+
+	uniterAPI, err := uniter.NewUniterAPI(
+		s.State,
+		s.resources,
+		s.authorizer,
+	)
+	c.Assert(err, jc.ErrorIsNil)
+	s.uniter = uniterAPI
+}
+
+var (
+	charmStatusWaiting = params.GoalStateStatus{
+		Status: "waiting",
+	}
+	expectedUnitWordpress = params.UnitsGoalState{
+		"wordpress/0": charmStatusWaiting,
+	}
+	expectedUnitMysql = params.UnitsGoalState{
+		"mysql/0": charmStatusWaiting,
+	}
+	expected2UnitsMysql = params.UnitsGoalState{
+		"mysql/0": charmStatusWaiting,
+		"mysql/1": charmStatusWaiting,
+	}
+)
+
+// TestGoalStatesNoRelation test single application with single unit
+func (s *uniterGoalStateSuite) TestGoalStatesNoRelation(c *gc.C) {
+	args := params.Entities{Entities: []params.Entity{
+		{Tag: "unit-mysql-0"},
+		{Tag: "unit-wordpress-0"},
+	}}
+	expected := params.GoalStateResults{
+		Results: []params.GoalStateResult{
+			{
+				Result: &params.GoalState{
+					Units: expectedUnitMysql,
+				},
+			}, {
+				Error: apiservertesting.ErrUnauthorized,
+			},
+		},
+	}
+	testGoalStates(c, s.uniter, args, expected)
+}
+
+// TestGoalStatesNoRelationTwoUnits adds a new unit to mysql application.
+func (s *uniterGoalStateSuite) TestGoalStatesNoRelationTwoUnits(c *gc.C) {
```

Comment:
The name of this test is not quite right. it implies we are testing a data model with 2 units but we're not. The second arg is simply to check that permissions are properly processed.
The test should be called something like TestPeerUnitsNoRelation()
Note that because the entire suite is GoalStateSuite, having GoalState in each test name is redundant.